NullPointerException when trying to send to an invalid email address

e.g.

```
2019-08-28 13:15:00,712 [pool-4-thread-1 - - -] INFO  rems.poller.email - sending email: {:to-user "RDapplicant2@funet.fi", :subject "Added as a member of an application (2019/10)"
, :body "Dear RDapplicant2@funet.fi,\n\nYou've been added as a member of application 2019/10.\n\nView application: https://rems2demo.csc.fi/#/application/10\n\nPlease don't reply to this automatically generated message.", :from "rems@csc.fi", :to "RDapplicant2.test@test_example.org"}
2019-08-28 13:15:00,875 [pool-4-thread-1 - - -] ERROR rems.scheduler - Internal error {:poller :rems.poller.email/poller,
 :event
 {:event/type :application.event/member-added,
  :event/time #clj-time/date-time "2019-08-15T10:24:11.451Z",
  :event/actor "RDapprover1@funet.fi",
  :application/id 10,
  :application/member {:userid "RDapplicant2@funet.fi"},
  :event/id 52}}

clojure.lang.ExceptionInfo: :rems.poller.email/poller failed to process event {:event/type :application.event/member-added, :event/time #clj-time/date-time "2019-08-15T10:24:11.451Z", :event/actor "RDapprover1@funet.fi", :application/id 10, :application/member {:userid "RDapplicant2@funet.fi"}, :event/id 52}
        at rems.poller.common$run_event_poller$fn__30083.invoke(common.clj:32)
        at rems.poller.common$run_event_poller.invokeStatic(common.clj:26)
        at rems.poller.common$run_event_poller.invoke(common.clj:18)
        at rems.poller.email$run.invokeStatic(email.clj:227)
        at rems.poller.email$run.invoke(email.clj:226)
        at rems.scheduler$start_BANG_$task__25394.invoke(scheduler.clj:11)
        at clojure.lang.AFn.run(AFn.java:22)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException: null
        at com.sun.mail.smtp.SMTPTransport.sendMessage(SMTPTransport.java:1204)
        at postal.smtp$smtp_send_STAR_.invokeStatic(smtp.clj:36)
        at postal.smtp$smtp_send_STAR_.invoke(smtp.clj:29)
        at postal.smtp$smtp_send.invokeStatic(smtp.clj:58)
        at postal.smtp$smtp_send.doInvoke(smtp.clj:39)
        at clojure.lang.RestFn.invoke(RestFn.java:423)
        at postal.core$send_message.invokeStatic(core.clj:36)
        at postal.core$send_message.invoke(core.clj:29)
```

This is caused by the invalid domain `test_example.org`.

Postal bug: https://github.com/drewr/postal/issues/62

--- a/src/clj/rems/poller/email.clj
+++ b/src/clj/rems/poller/email.clj
@@ -206,16 +206,16 @@
 
 (defn send-email! [email-spec]
   (let [host (:smtp-host env)
-        port (:smtp-port env)]
+        port (:smtp-port env)
+        email (assoc email-spec
+                     :from (:mail-from env)
+                     :to (or (:to email-spec)
+                             (util/get-user-mail
+                              (users/get-user-attributes
+                               (:to-user email-spec)))))]
     (if (not (and host port))
-      (log/info "pretending to send email:" (pr-str email-spec))
-      (let [email (assoc email-spec
-                         :from (:mail-from env)
-                         :body (:body email-spec)
-                         :to (or (:to email-spec)
-                                 (util/get-user-mail
-                                  (users/get-user-attributes
-                                   (:to-user email-spec)))))]
+      (log/info "pretending to send email:" (pr-str email))
+      (do
         ;; TODO check that :to is set
         (log/info "sending email:" (pr-str email))
         (try

--- a/src/clj/rems/poller/email.clj
+++ b/src/clj/rems/poller/email.clj
@@ -233,20 +233,19 @@
                               (users/get-user-attributes
                                (:to-user email-spec)))))
         to-error (validate-address (:to email))]
+    (log/info "sending email:" (pr-str email))
     (cond
       to-error
-      (log/warn "failed address validation: " to-error)
+      (log/warn "failed address validation, skipping message: " to-error)
 
       (not (and host port))
-      (log/info "pretending to send email:" (pr-str email))
+      (log/info "no smtp server configured, only pretending to send email")
 
       :else
-      (do
-        (log/info "sending email:" (pr-str email))
-        (try
-          (postal/send-message {:host host :port port} email)
-          (catch com.sun.mail.smtp.SMTPAddressFailedException e ; email address does not exist
-            (log/warn e "failed sending email, skipping:" (pr-str email))))))))
+      (try
+        (postal/send-message {:host host :port port} email)
+        (catch com.sun.mail.smtp.SMTPAddressFailedException e ; email address does not exist
+          (log/warn e "failed sending email, skipping:" (pr-str email)))))))
 
 (defn run []
   (common/run-event-poller ::poller (fn [event]

--- a/src/clj/rems/poller/email.clj
+++ b/src/clj/rems/poller/email.clj
@@ -15,7 +15,8 @@
             [rems.scheduler :as scheduler]
             [rems.text :refer [text text-format with-language]]
             [rems.util :as util])
-  (:import [org.joda.time Duration]))
+  (:import [javax.mail.internet AddressException InternetAddress]
+           [org.joda.time Duration]))
 
 ;;; Mapping events to emails
 
@@ -204,6 +205,24 @@
         last-id (:event/id (last events))]
     (common/set-poller-state! ::poller {:last-processed-event-id last-id})))
 
+(defn- validate-address
+  "Returns nil for a valid email address, string message for an invalid one."
+  [email]
+  (try
+    (InternetAddress. email)
+    nil
+    (catch Throwable t
+      (str "Invalid address "
+           (pr-str email)
+           ": "
+           (.getMessage t)))))
+
+(deftest test-validate-address
+  (is (nil? (validate-address "valid@example.com")))
+  (is (string? (validate-address "")))
+  (is (string? (validate-address nil)))
+  (is (string? (validate-address "test@test_example.com"))))
+
 (defn send-email! [email-spec]
   (let [host (:smtp-host env)
         port (:smtp-port env)
@@ -212,11 +231,17 @@
                      :to (or (:to email-spec)
                              (util/get-user-mail
                               (users/get-user-attributes
-                               (:to-user email-spec)))))]
-    (if (not (and host port))
+                               (:to-user email-spec)))))
+        to-error (validate-address (:to email))]
+    (cond
+      to-error
+      (log/warn "failed address validation: " to-error)
+
+      (not (and host port))
       (log/info "pretending to send email:" (pr-str email))
+
+      :else
       (do
-        ;; TODO check that :to is set
         (log/info "sending email:" (pr-str email))
         (try
           (postal/send-message {:host host :port port} email)

